packages/neuron-wallet/src/services/wallets.ts

```diff
@@ -88,6 +149,77 @@ export class FileKeystoreWallet implements Wallet {
   keystoreFileName = () => {
     return `${this.id}.json`
   }
+
+  public checkAndGenerateAddresses = async (
+    isImporting: boolean = false,
+    receivingAddressCount: number = DefaultAddressNumber.Receiving,
+    changeAddressCount: number = DefaultAddressNumber.Change
+  ): Promise<AddressInterface[] | undefined> => {
+      return await AddressService.generateAndSaveForExtendedKey(
+        this.id,
+        this.accountExtendedPublicKey(),
+        isImporting,
+        receivingAddressCount,
+        changeAddressCount
+      )
+  }
+
+  public getNextAddressByWalletId = async (): Promise<AddressInterface | undefined> => {
+    return AddressService.getNextUnusedAddressByWalletId(this.id)
+  }
+
+  public getNextChangeAddressByWalletId = async (): Promise<AddressInterface | undefined> => {
+    return AddressService.getNextUnusedChangeAddressByWalletId(this.id)
+  }
+
+  public getNextReceivingAddressesByWalletId = async (): Promise<AddressInterface[]> => {
+    return AddressService.getUnusedReceivingAddressesByWalletId(this.id)
+  }
+}
+
+export class HardwareWallet extends Wallet {
+  public isHardware = (): boolean => {
+    return true
+  }
+
+  static fromJSON = (json: WalletProperties) => {
+    return new HardwareWallet(json)
+  }
+
+  public getDeviceInfo = (): DeviceProperties => {
+    return this.device!
+  }
+
+  public checkAndGenerateAddresses = async (): Promise<AddressInterface[] | undefined> => {
+    const {publicKey, addressType, addressIndex} = this.getDeviceInfo()
+    const address = await AddressService.generateAndSaveForPublicKey(
+      this.id,
+      publicKey,
+      addressType,
+      addressIndex
+    )
+
+    if (address) {
+      return [address]
+    }
+  }
+
+  public getNextAddressByWalletId = async (): Promise<AddressInterface | undefined> => {
+    return AddressService.getFirstAddressByWalletId(this.id)
```

Comment:
I find it a bit odd that the function is named `getNextAddressByWalletId`, since it doesn't take a `walletId` argument. Maybe `getNextAddress` is enough?

Reply:
Ah right. The result of copy and paste. `getNextAddress` is a better naming.